barber.h

```diff
@@ -1,27 +1,32 @@
 #ifndef BARBER_H
 #define BARBER_H
 
-#include <iostream>
 #include <thread>
-#include <chrono>
-
-using namespace std;
+#include <future>
 
 class Shop;
+class Semaphore;
 
 class Barber
 {
 public:
     Barber(Shop* shop);
-    void terminate();
+    void joinThread();
+    bool stopRequested();
+    void stop();
 
 private:
     std::thread mBarberThread;
-    bool mAlive;
+
+    std::promise<void> exitSignal;
+    std::future<void> futureObj;
+
     Shop* pBelongsToShop;
+    Semaphore* mBarberNotifier;
+    Semaphore* mCustomersNotifier;
```

Comment:
Is there any benefit to keep here both the shop and the shop's semaphores? You can always get the semaphores from the shop.

Reply:
Done :ok: